usaspending_api/etl/management/commands/combine_universal_transaction_matview_chunks.py

```diff
@@ -0,0 +1,200 @@
+
+        logger.info(f"Chunk Count: {chunk_count}")
+
+        with Timer("Recreating table"):
+            self.recreate_matview()
+
+        with Timer("Inserting data into table"):
+            self.insert_matview_data(chunk_count)
+
+        with Timer("Creating table indexes"):
+            self.create_indexes()
+
+        with Timer("Swapping Tables/Indexes"):
+            self.swap_matviews()
+
+        if not options["keep_old_data"]:
+            self.remove_old_data(chunk_count)
+
+        if options["analyze"]:
+            with Timer("Analyzing Table"):
+                self.analyze_matview()
+
+        with Timer("Granting Table Permissions"):
+            self.grant_matview_permissions()
+
+    def recreate_matview(self):
+        with connection.cursor() as cursor:
+            # This table was previously a Matview, so the DROP may fail
+            try:
+                cursor.execute(RECREATE_TABLE_SQL.format(old_object_type="TABLE"))
+            except Exception as e:
+                if "is not a table" in str(e):
+                    logger.info(
+                        "universal_transaction_matview_temp existed, but not as table. Trying to drop as Matview"
+                    )
+                    cursor.execute(RECREATE_TABLE_SQL.format(old_object_type="MATERIALIZED VIEW"))
+
+    def insert_matview_data(self, chunk_count):
+        loop = asyncio.new_event_loop()
+        tasks = []
+
+        for current_chunk in range(0, chunk_count):
+            tasks.append(
+                asyncio.ensure_future(
+                    async_run_creates(
+                        INSERT_INTO_TABLE_SQL.format(current_chunk=current_chunk),
+                        wrapper=Timer(f"Insert into table {current_chunk}"),
+                    ),
+                    loop=loop,
+                )
+            )
+
+        loop.run_until_complete(asyncio.gather(*tasks))
+        loop.close()
+
+    def create_indexes(self):
+        loop = asyncio.new_event_loop()
+        tasks = []
+        i = 0
+        for sql in sqlparse.split(TABLE_INDEX_SQL):
+            tasks.append(
+                asyncio.ensure_future(async_run_creates(sql, wrapper=Timer(f"Creating Index {i}"),), loop=loop,)
+            )
+            i += 1
+
+        loop.run_until_complete(asyncio.gather(*tasks))
+        loop.close()
+
+    def swap_matviews(self):
+        with connection.cursor() as cursor:
+            # This table was previously a Matview, so the DROP may fail
+            try:
+                cursor.execute(SWAP_TABLES_SQL.format(old_object_type="TABLE"))
+            except Exception as e:
+                if "is not a table" in str(e):
+                    logger.info(
+                        "universal_transaction_matview_temp existed, but not as table. Trying to drop as Matview"
+                    )
+                    cursor.execute(SWAP_TABLES_SQL.format(old_object_type="MATERIALIZED VIEW"))
+
+    def remove_old_data(self, chunk_count):
```

Comment:
This will get called by default unless someone supplies `--keep-old-data`

The removal of table with suffix `_old` is fine. But this:
`REFRESH MATERIALIZED VIEW universal_transaction_matview_{current_chunk} WITH NO DATA;`

Getting called will _empty_ out each of the matview chunks. Meaning the next night, they need to be created from scratch again rather than updated via a `REFRESH` (which under the hood compares old rows to new rows).

Is that what we want ☝️ ??
- I would have though that doing a refresh from night to night rather than a recreate on the matviews would save on execution time of "re-hydrating" the matview chunks with new data.

Reply:
> Is that what we want ☝️ ??
@kwhickey yes. Why not? @collinwr said the times of the mativew creation was very similar to a refresh. Are there more than these three options?

1. Leave the "chunked" matviews in-place during the day, consuming approx 120GB but useless for any queries
2. Refresh with no data which recovered the storage and leaves the matviews around for simpler pipeline operations
3. Drop the matviews afterward, which would be closer to "created from scratch".


@collinwr currently the refresh SQL will fail. A materialized view with no data cannot be concurrently refreshed. Additionally, even with data a concurrent refresh requires a unique index on the matview

Reply:
I'm not sure we need to perform the refresh concurrently because the matviews are not live (being used by the application). They are only used to insert data into the table. Is there another reason to do the refresh concurrently?

Reply:
@collinwr we don't need concurrent refreshes. But unless I am mistaken the SQL being stored in `*__refresh.sql` for the chunked materialized views is
```sql
REFRESH MATERIALIZED VIEW CONCURRENTLY universal_transaction_matview_9 WITH DATA;
VACUUM ANALYZE VERBOSE universal_transaction_matview_9;
```

Reply:
We should chat about this. If we're recreating the mativiews from scratch (going from NO DATA to DATA everyday) ... why are they matviews? The chunks should just be tables that are truncated and inserted into created from scratch each day. 

Matviews have the benefit of a REFRESH to only update _part_ of them. And we're not using that benefit.

Reply:
Open and available to chat.

> why are they matviews

Why not? Matviews have several benefits, and theses scripts _are_ leveraging matview capabilities. I am curious to hear why recovering the data is concerning to you